### PR TITLE
gh-120600: Make Py_TYPE() opaque in limited C API 3.14

### PR DESCRIPTION
**Pull request by vstinner**
In the limited C API 3.14 and newer, Py_TYPE() and Py_SET_TYPE() are now implemented as opaque function calls to hide implementation details.

* Issue: gh-120600

----
📚 Documentation preview 📚: https://cpython-previews--120601.org.readthedocs.build/

### CODE REVIEW

**Comment by vstinner** [comment]
cc @encukou @serhiy-storchaka @zooba

**Review by serhiy-storchaka — COMMENTED** [review]
There is `PyObject_Type()`.

**Comment by vstinner** [comment]
> There is PyObject_Type().

I don't understand your remark. Are you suggesting to reimplement Py_TYPE() on top of PyObject_Type()? Py_TYPE() returns a borrowed reference, PyObject_Type() returns a strong reference.

**Comment by serhiy-storchaka** [comment]
Py_TYPE was defined as a macro, it depends on the internal structure of PyObject. If you need to use the limited C API, use PyObject_Type() instead of Py_TYPE().

**Comment by vstinner** [comment]
> Py_TYPE was defined as a macro, it depends on the internal structure of PyObject. If you need to use the limited C API, use PyObject_Type() instead of Py_TYPE().

Py_TYPE() is very commonly used in C extensions. Having to switch to PyObject_Type() is a lot of work. I don't want to force C extensions maintainers to do that. I only want to hide the implementation details (access to `PyObject.ob_type` member, maybe with an atomic operation in Free Threading).

It's not convenient to have to release a reference when using PyObject_Type(), compared to Py_TYPE().

**Comment by serhiy-storchaka** [comment]
Has not C API returning a borrowed reference been frowned upon lately? Strictly speaking, Py_TYPE() is incompatible with Free Threading.

If you want to use the limited C API, use PyObject_Type(). If you prefer performance and convenience, use Py_TYPE() and the GIL.

**Comment by vstinner** [comment]
I don't get it. The stable ABI is guaranteed to be stable. Py_TYPE() is part of the stable ABI. I'm not asking to add it. Only to change its implementation.

**Comment by serhiy-storchaka** [comment]
Than what is the point of changing `Py_TYPE()`? If you do not use the limited C API, you rely on implementation details. If you do not want to rely on them, use `PyObject_Type()`.

File: Misc/stable_abi.toml [hunk 1]
```diff
@@ -2507,3 +2507,5 @@
     added = '3.13'
 [function.PyEval_GetFrameLocals]
     added = '3.13'
+[function.Py_TYPE]
+    added = '3.14'
```
Comment (encukou) [inline Misc/stable_abi.toml:2512]: Please leave a blank line before the first 3.14 entry :)
```suggestion

[function.Py_TYPE]
    added = '3.14'
```

**Review by encukou — APPROVED** [review]
+1 from me. Even if we can't change the current version of the stable ABI, moving to functions makes sense.
Exposed structs are [the one biggest issue with `abi4`](https://github.com/capi-workgroup/problems/issues/4#issuecomment-1540917809); if there's ever a new version of a stable ABI, it probably won't have them.
It would be great if modules compiled for the *next* stable ABI are compatible with 3.14, and at the same time stay *API*-compatible.

**Review by serhiy-storchaka — APPROVED** [review]

**Pull request merged**